### ISSUE TITLE
Using enums: "Type Error. Expecting an integral type"

### ISSUE THREAD

**Issue by randaz81** [swig/swig#2796]
with swigwin 4..2.0:
```
typedef std::int32_t mytype1;
typedef int mytype2;

enum myEnum
{
   CASE0 = 10,                  //OK
   CASE1 = (int)(10),           //OK
   CASE2 = (std::int32_t)(10),  //OK
   CASE3 = (mytype1)(10),       //FAILS
   CASE4 = (mytype2)(10)        //OK
};
```
It looks like a failure in parsing the typedef of a type which is a typedef too.
The error was not present with swigwin 4.0.2

**Comment by ojwb** [comment]
(Your example was missing a `;` after the closing `}` - I've added one so it's actually valid C++.)

`git bisect` points to 3a6d12475141cbace350d1ef7cddd38232c2a7eb.

I wondered if this would be the same underlying problem as ec3541e988d72be4241a12bfa96397a18f73a3e0 (since the type `std::int32_t` is unknown to SWIG here) but that commit didn't fix this example.

Not specific to C++ as removing `std::` and processing without `-c++` gives the same error.

**Comment by ojwb** [comment]
This patch fixes the error:

```
diff --git a/Source/CParse/parser.y b/Source/CParse/parser.y
index ba65be63..09c05924 100644
--- a/Source/CParse/parser.y
+++ b/Source/CParse/parser.y
@@ -6535,7 +6535,10 @@ expr           : valexpr
 		   } else {
 		     SwigType *type = Getattr(n, "type");
 		     if (type) {
-		       $$.type = SwigType_type(type);
+		       int type_code = SwigType_type(type);
+		       if (type_code != T_USER) {
+			 $$.type = type_code;
+		       }
 		     }
 		   }
 		 }
```

However I'm not going to merge this as it seems the wrong way to address this - `T_USER` is the most appropriate type code here, and it's whatever ends up consuming it that needs adjusting to handle that we now set the type code better (in SWIG 4.1.1 and below we always set it to `T_INT` in this grammar rule, which was sometimes actually right by luck (like a stopped clock).

I'll track down a better fix and merge it for 4.2.1.  Possibly the answer is to relax the check that reports the error to accept `T_USER` since that could be a typedef to an unknown type like in this case.

I'm not entirely clear why `(std::int32_t)(10)` seems to work though.  Perhaps because SWIG knows nothing about `std::int32_t` but it does know `mytype1` is a typedef.  I think that's worth investigating too.

**Comment by ojwb** [comment]
```
$ cat allcases.i
#ifdef SWIG
%module x
#endif
typedef int mytype0;
typedef mytype0 mytype1;
#ifndef SWIG
typedef int unknown_to_swig_type;
#endif
typedef unknown_to_swig_type mytype2;

enum myEnum
{
   CASE0A = (mytype0)10,  //OK
   CASE0B = mytype0(10),  //OK
   CASE0C = static_cast<mytype0>(10), // OK
   CASE1A = (mytype1)10,       //FAILS
   CASE1B = mytype1(10),        //OK
   CASE1C = static_cast<mytype1>(10), // OK
   CASE2A = (mytype2)10,       //FAILS
   CASE2B = mytype2(10),        //OK
   CASE2C = static_cast<mytype2>(10), // OK
   CASE3A = (unknown_to_swig_type)10,       //OK
   CASE3B = unknown_to_swig_type(10),        //OK
   CASE3C = static_cast<unknown_to_swig_type>(10), // OK
};
$ g++ -Wall -W -O2 -c -xc++ allcases.i 
$ ../preinst-swig -python -c++ allcases.i
allcases.i:16: Error: Type error. Expecting an integral type
allcases.i:19: Error: Type error. Expecting an integral type
$
```

Debugging it seemed to me something like this should fix at least the first error above:

```
diff --git a/Source/CParse/parser.y b/Source/CParse/parser.y
index ba65be63..83038aea 100644
--- a/Source/CParse/parser.y
+++ b/Source/CParse/parser.y
@@ -6535,7 +6535,10 @@ expr           : valexpr
 		   } else {
 		     SwigType *type = Getattr(n, "type");
 		     if (type) {
+		       SwigType *true_type = SwigType_typedef_resolve_all(type);
 		       $$.type = SwigType_type(type);
+		       $$.type = SwigType_type(true_type);
+		       Delete(true_type);
 		     }
 		   }
 		 }
```

However `SwigType_typedef_resolve_all()` doesn't resolve `mytype0` to `int` (or `mytype2` to `unknown_to_swig_type`).  I'm not clear why not, but I wonder if the current scope it seems to use isn't properly set up at this point.  There don't seem to be any existing calls to resolve typedefs as we parse, but trying to delay this until later doesn't seem trivial.

@wsfulton Is it possible to resolve a typedef in the parser?

If not, I wonder if we're best to just add `T_USER` to the list of type codes which don't trigger this type error.

**Comment by wsfulton** [comment]
Oh boy, I think this is a can of worms. Typedef resolution is done in the 2nd stage of parsing not the first, so trying to fully resolve the type seems to be going against the entire design and will need re-implementing a lot of what is in Typepass class in the parser. I think the only successful typedef resolution in the parser is for template types (and that is of course not bug-free and it is a lot of code).

I think it is probably best given there were lots of commits around this attempt at type deduction in the parser and a fix is needed quickly for 4.2.1, is maybe to remove that error in etype and replace the non-numeric types with T_INT (mostly like how it used to be). Or probably better accept T_USER as you suggested, and then defer checking to typepass. Although on second thoughts, how one deconstructs an expression into its components is not something we seem to handle. Expressions have a long way to go in SWIG to be fully supported I think, it's something I avoid touching, but is somewhat lacking. Maybe the closest behaviour to what we had before was to convert T_USER into T_INT in etype in the parser.

**Comment by ojwb** [comment]
> how one deconstructs an expression into its components is not something we seem to handle

I think we mostly do (unless you mean something different to what I think you mean by that).  The parser effectively deconstructs each expression and then it gets reassembled (and the type code of the expression tracked) through each of the expression-related grammar rules.

There are admittedly some awkward corner cases where the parse of something is unexpected (at least to me) - e.g. the expression `(6)*7` here:

```
enum myEnum
{
   E = (6)*7
};
```

is parsed by this rule which seems to be intended for casts like `(int)42`:

```
/* A few common casting operations */

               | LPAREN expr RPAREN expr %prec CAST {
```

The second `expr` being `*7` where `*` is the unary dereference operator, so SWIG parses this as "7 dereferenced then cast to type 6" - the type deduction in this case is currently a bit of a bodge unfortunately.  It's correct when this is multiplication, but something like `(short)42` is treated as having type `int`.

We can probably resolve this (and any other cases of this sort - this is the only one I'm aware of currently, but there are a lot of hacks in the grammar so there may be others) once we switch to generating a GLR parser.

> Maybe the closest behaviour to what we had before was to convert T_USER into T_INT in etype in the parser.

Indeed.  Given we can't easily resolve typedefs in the parser, I'll go for that.

**Comment by wsfulton** [comment]
> > how one deconstructs an expression into its components is not something we seem to handle
> 
> I think we mostly do (unless you mean something different to what I think you mean by that). The parser effectively deconstructs each expression and then it gets reassembled (and the type code of the expression tracked) through each of the expression-related grammar rules.

I see my comment is confusing! What I meant is deconstructing from the parse tree, so for example, CASE2C above. The parser deconstructs the expression, but them promptly reassembles it so that in the parse tree, all we see is
```
      | enumvalue    - "static_cast< mytype1 >(10)"
```
It isn't even a `SwigType`, it's just a pure string, mostly as parsed. As expressions can contain type info, we really need a way to break the expression down into its components after the initial parser stage so that the types can be fully deduced/resolved. In the case above, `mytype1`, is quite simple and wouldn't be too challenging to fully resolve, but imagine an expression using templates and complicated template parameters. SWIG can actually handle these, but they are done later on after the initial parse stage.

**Comment by ojwb** [comment]
Thanks for clarifying.  We do also have the integer type code, though for `static_cast< mytype1 >(10)` it's probably just `T_UNKNOWN`.  It would be better if we tracked types through the parsing of expressions as `SwigType` rather than the rather limited set of integer type codes (`T_INT`, etc), though if we can't resolve typedefs until later then for cases such as `CASE2C` that's going to either be tricky or have to report "unknown type" when we could do better (we would need to be able to report "unknown type" for `CASE3A`, `CASE3B` and `CASE3C` as SWIG genuinely doesn't know the type there).

It looks like we can't even tell if something is a typedef at this stage of parsing, since `SwigType_istypedef()` works by trying to resolve typedefs.

Prodding at this, I found another regression - `CASE4 = (6)&7` still gives `Error: Type error. Expecting an integral type` because we get type code `T_POINTER`, but worked with SWIG 4.1.0 (and I'd expect 4.1.1 but I don't have that to hand to test).  Probably the simplest fix is to switch unary `&` to just give `T_UNKNOWN` which I'd think is unlikely to affect many real-world cases (`T_POINTER` isn't actually going to help us deduce a type anyway; `T_STRING` for `&'a'` and similarly for `T_WSTRING` could, but the cases where it does seem quite artificial).  I'll sort out a fix; reopening to track that.
